src/browser.cc

```diff
@@ -0,0 +1,369 @@
+        {  // TABLE, TR.
+          TABLE table({{"border", "0"}, {"align", "center"}, {"cellpadding", "8"}});
+          TR r({{"align", "center"}});
+          if (one_based_index > 1) {
+            TD d;
+            A a({{"href", FLAGS_route + "?id=" + ToString(one_based_index - 1) + "&html=yes"}});
+            TEXT("Previous insight");
+          }
+          if (one_based_index < input.insight.size()) {
+            TD d;
+            A a({{"href", FLAGS_route + "?id=" + ToString(one_based_index + 1) + "&html=yes"}});
+            TEXT("Next insight");
+          }
+        }  // TABLE, TR.
+        input.insight[one_based_index - 1]->RenderHTML(input.feature);
+        r(html_scope.AsString(), HTTPResponseCode.OK, "text/html");
+      } else {
+        r(InsightResponse(input, one_based_index - 1));
+      }
+    } else if (r.url.query["id"] == "all") {
+      r(input.insight, "insights");
+    } else if (r.url.query["id"] == "everything") {
+      r(input, "everything");
+    } else {
+      r(TopLevelResponse(input.insight.size()));
+    }
+  });
+
+  WaitableAtomic<SmartSessionInfoMap> sessions;
+
+  HTTP(FLAGS_port).Register(FLAGS_route + "smart", [&input, &sessions](Request r) {
+    const bool as_html = !r.url.query["html"].empty();
+    const std::string& id = r.url.query[FLAGS_id_key];
+    const std::string& action = r.url.query["action"];
+    if (id.empty()) {
+      // Create new user session ID and redirect to it.
+      const std::string fresh_id = RandomString();
+      const std::string P = FLAGS_output_uri_prefix + FLAGS_route + "smart?" + FLAGS_id_key + '=' + fresh_id +
+                            "&html=" + (as_html ? "yes" : "");
+      r("", HTTPResponseCode.Found, "text/html", HTTPHeaders({{"Location", P}}));
+    } else {
+      // Smart session browsing.
+      SmartInsightResponse payload;
+      sessions.MutableUse([&](SmartSessionInfoMap& map) {
+        auto& info = map[id];
+        info.TakeAction(input, action, payload, id, as_html);
+        if (info) {
+          payload.done = false;
+          payload.insight.Prepare(input, info.current_insight_index);
+        } else {
+          payload.done = true;
+        }
+        payload.sessions = map;
+      });
+      if (!as_html) {
+        r(payload, "smart_insight");
+      } else {
+        using namespace html;
+        HTML html_scope;
+        {  // HEAD.
+          HEAD head;
+          TITLE("Insights Visualization Alpha");
+        }
+        if (payload.done) {
+          TEXT("<h1>You are awesome!</h1>");
+          {
+            A a({{"href", FLAGS_output_uri_prefix + FLAGS_route + "smart?html=yes"}});
+            TEXT("Start over!");
+          }
+        } else {
+          for (const auto& nav : payload.navigation) {
+            TEXT("<p align=center>");
+            A a({{"href", nav.uri}});
+            TEXT(nav.text);
+            TEXT("</p>");
+          }
+          TEXT("<hr>");
+          {
+            TEXT("<p align=center>");
```

Comment:
:-1: my bad bro.

Won't be fixing C++/HTML logic though -- it's very temporary here.